### DISCUSSION TITLE
Building for Windows on Mac? Also, Mac building without XCode?

### DISCUSSION THREAD

**Discussion by SasoriJr** [Kode/Kha#1431] [Q&A]
Hi everyone, first post here. I've been using Kha for a while and I'll be delivering a demo of my game in the next weeks. As I'm using an old iMac 2011 (High Sierra), I'm able to build for Mac (with XCode) and Linux, but not Windows. I remember I could build for all 3 when I was using monogame/dotnet (for Windows and for Mac (without XCode)). Would that be possible to do with Kha? Thank you :)

**Comment by RobDangerous** [comment]
Sorry, there's no functionality in Kha for that. And even in theory it's only somewhat possible - for Windows you need to compile shaders for Direct3D and that compiler is not available for Non-Windows. You could do an OpenGL or Vulkan-build but that's not as compatible. Or you could just not pre-compile the shaders but that's also not great. So whatever you do, cross-compiling for Windows will result in a build which is in some way compromised. Doing something without Xcode is also not an option because Xcode is what gets you the C/C++-compiler in the first place.

**Reply by SasoriJr** [reply]
Hey, thanks for explaining, I can see why now. I'm going to set up a VirtualBox session so I can build it.

**Reply by RobDangerous** [reply]
GitHub actions can also be a very convenient option.

**Reply by SasoriJr** [reply]
You're right. Thanks for the advice.

**Comment by MetalMaxMX** [comment]
Bit of a follow-up question to this since it is fairly relevant to the discussion. You mentioned that making it possible to build from macOS would be difficult because any Windows-build would be compromised by the lack of Direct3D support on those platforms. 

I come from a Linux background so I am curious to know something. I have seen some projects where they support the topic of cross-building with a compiler known as MinGW (or MinGW-w64) and this compiler is available both on Linux and Windows. If Kha supported being able to generate makefiles or builds for those compilers, would it be possible to get Windows builds? (be it OpenGL using GLSL shaders or Vulkan builds on Windows)

**Reply by RobDangerous** [reply]
Mingw is not really a compiler itself, it's a version of GCC and a bunch of Windows libraries which have been adjusted to work with that, if I understand it correctly. Somebody tried that recently (from Windows though) and the result was that the created binaries crashed on startup on some computers - then they switched to msvc and everything was fine. I don't know the exact reasons and I'm not terribly interested in spending time to find out. In addition to that the problems I mentioned in the previous answer also hold true for mingw in addition to mingw's own problems. If you want to build from Linux or macOS better give clang a try which directly supports regular Windows libraries nowadays.

**Reply by MetalMaxMX** [reply]
Does Clang comes with all the libraries and tooling needed to build from Linux/macOS to Windows?

**Reply by RobDangerous** [reply]
I don't think you need any tooling aside from the clang binary itself. But it does not come with any libs, you will have to put some work in. Or maybe somebody has a nice package somewhere, don't know.

**Reply by MetalMaxMX** [reply]
I have seen some tutorials and all in how they approach cross-building with Clang and it seems to be a mix of Dockerfiles with strategic packages laid around in several locations and so? 

I think at this point VirtualBox or CIs might be better for Kha here. Still, thanks for answering! (Didn't know MinGW was that problematic for your users admittedly) 

(I have also heard of vcpkg and doing triplets with that in other projects, though I heard that's also problematic in its own way I suppose?)

**Reply by RobDangerous** [reply]
Mingw is a hacky thing that usually works - but good luck figuring out what's wrong when it doesn't.
I don't see a reason why clang would require dockerfiles. Build a nice and easy to use package with that and become famous.

**Reply by MetalMaxMX** [reply]
It was a tutorial where someone was trying to get a sysroot from an existing compiler installation, which is often done with Dockerfiles (another approach was making yet another dockfile containing Rust's LLVM to handle multiple-targets as well) and that's why I thought it was necessary to get dockerfiles to work with a cross-compilation setup on LLVM/Clang. 

I might have to try a few things (like copying libs from a Clang installation on Windows to Linux and see if by adding a triplet I can get something working? Since Linux already comes with headers and libs for most packages, perhaps that approach could work?), would've loved to see any more Kha stories surrounding this setup. Somewhere who has documented some form of MinGW or Clang cross-compilation setup on Kha since it doesn't seem to be explored ground?

**Reply by RobDangerous** [reply]
Be the change you want to see in the world.

**Reply by MetalMaxMX** [reply]
Well, I'll give you that! Sometimes the best things are only gotten when one does it themselves.
